### PR TITLE
Improve date and datetime formatting

### PR DESCRIPTION
**Pull request by Joonalai**
## Description
This PR allows formatting date and datetime field values based on the selected locale.

This improves the solution made in https://github.com/qgis/QGIS/pull/45656 according to the [comments](https://github.com/qgis/QGIS/pull/46193#issuecomment-977891977) made by @m-kuhn. A separate PR for backporting this solution to 3.22 will be made after this is merged. 

WIP:
- [x] some minor details has to be discussed before merging.

### CODE REVIEW

**Comment by Joonalai** [comment]
@m-kuhn I am not sure about the `setLocale` setter in `QgsApplication`. I could not use that setter to set the locale here https://github.com/qgis/QGIS/blob/ef91c0e046bc570c3b4db1c788b92644e68f6b6c/src/app/main.cpp#L1017 or here https://github.com/qgis/QGIS/blob/ef91c0e046bc570c3b4db1c788b92644e68f6b6c/src/app/main.cpp#L1030 since `QgsApplication`'s constructor uses `sTranslation`.  Should I maybe just emit the signal here https://github.com/qgis/QGIS/blob/ef91c0e046bc570c3b4db1c788b92644e68f6b6c/src/app/main.cpp#L1038?

**Comment by m-kuhn** [comment]
> since QgsApplication's constructor uses sTranslation. Should I maybe just emit the signal here

That sounds good, yes. The setter can't really be used on runtime anyway (a restart is always required for the UI locale changes), so if that works, perfect.

**Comment by Joonalai** [comment]
@m-kuhn emitting that signal in tests seems a bit weird, so maybe we go with the setter after all. It could be handy in tests and plugins after all.

**Comment by m-kuhn** [comment]
also good, let's make sure the limitations are documented

File: src/core/qgsapplication.h [hunk 1]
```diff
@@ -449,6 +449,12 @@ class CORE_EXPORT QgsApplication : public QApplication
      */
     static QString locale();
 
+    /**
+     * Sets the QGIS locale.
```
Comment (m-kuhn) [inline src/core/qgsapplication.h:453]: This needs some updates to clarify that this is mostly there for custom applications, whereas in QGIS this is internally triggered by the application already.

File: src/core/qgsapplication.h [hunk 2]
```diff
@@ -449,6 +449,12 @@ class CORE_EXPORT QgsApplication : public QApplication
...
+     * \since QGIS 3.22
+     */
+    static void setLocale( QLocale locale );
```
Comment (m-kuhn) [inline src/core/qgsapplication.h:456]:
```suggestion
    static void setLocale( const QLocale &locale );
```

File: src/core/qgsapplication.h (hunk 1 above)
Comment (Joonalai) [reply src/core/qgsapplication.h]: Is it clear enough now?

File: src/core/qgsapplication.h [hunk 3]
```diff
@@ -1024,6 +1032,13 @@ class CORE_EXPORT QgsApplication : public QApplication
      */
     void requestForTranslatableObjects( QgsTranslationContext *translationContext );
 
+
+    /**
+     * Emitted when project locale has been changed.
```
Comment (m-kuhn) [inline src/core/qgsapplication.h:1037]: Forgot to ask, can you add `\since QGIS 3.24`

File: src/core/qgsapplication.h [hunk 4]
```diff
@@ -449,6 +449,14 @@ class CORE_EXPORT QgsApplication : public QApplication
      */
     static QString locale();
 
+    /**
+     * Sets the QGIS locale - used mainly by 3rd party apps and tests.
+     * In QGIS this is internally triggered by the application in startup.
+     *
+     * \since QGIS 3.22
```
Comment (m-kuhn) [inline src/core/qgsapplication.h:456]: 3.24

File: src/core/qgsapplication.h (hunk 1 above)
Comment (m-kuhn) [reply src/core/qgsapplication.h]: Yep, looks good

File: src/core/qgsapplication.h (hunk 3 above)
Comment (Joonalai) [reply src/core/qgsapplication.h:1037]: Yes, I will add the since clause.

File: src/core/qgsapplication.h (hunk 4 above)
Comment (Joonalai) [reply src/core/qgsapplication.h:456]: I tought that after backport this is available in 3.22. I will cherry-pick these commits (not the first one) to backport branch. What do you think, which one should we use?

Comment (m-kuhn) [reply src/core/qgsapplication.h:456]: I forgot that the backport was already discussed in the original PR.
Let's go for 3.22.2 or 3.22.3

Comment (Joonalai) [reply src/core/qgsapplication.h:456]: Alright, I will change those to 3.22.2.

**Review by m-kuhn — APPROVED** [review]

File: src/core/qgsapplication.cpp [hunk 5]
```diff
@@ -1310,6 +1313,12 @@ QString QgsApplication::locale()
   }
 }
 
+void QgsApplication::setLocale( const QLocale &locale )
+{
+  QLocale::setDefault( locale );
+  emit instance()->localeChanged();
```
Comment (m-kuhn) [inline src/core/qgsapplication.cpp:1319]: Nice choice to make this static :+1:

Comment (Joonalai) [reply src/core/qgsapplication.cpp:1319]: Thanks!

**Pull request merged**